Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,522 @@
+to learn more.
+
+### Generated Xcode projects
+
+In an Xcode world, folks checked in a project which contains an listing of
+files, build settings, and IDE state. When the project scales, the project file
+model breaks down. Auditing and code reviewing config changes in Xcode projects
+quickly becomes difficult. With Bazel, human readable `BUILD` files are the
+source of truth for build and Xcode configuration. Tools like
+[XCHammer](https://github.com/pinterest/xchammer), and
+[Tulsi](https://github.com/bazelbuild/tulsi) use an aspect to traverse the
+build graph and extract metadata required to generate a project.  These tools
+make it easier to manage the project and generate on demand - not needing to
+check it in.
+
+XCHammer provides a rule to [bazel build Xcode
+projects](ihttps://github.com/pinterest/xchammer#bazel-build-xcode-projects).
+Simply declare the rule with the project and Bazel build the target.
+```
+load("@xchammer_resources//:xcodeproject.bzl", "xcode_project")
+xcode_project(
+    name = "MyProject",
+    targets = [ "//ios-app:ios-app" ],
+    paths = [ "**" ],
+)
+```
+
+Like any other Bazel target, it's built from the command line
+```
+bazel build :MyProject
+```
+
+The rule definition for the `xcode_project` may be found in the [github
+repository](https://github.com/pinterest/xchammer/blob/master/BazelExtensions/xcodeproject.bzl).
+Simply put, the aspect traverses sources, and invokes the `xchammer` binary with
+a JSON file. Internally, XCHammer instantiates
+[XcodeGen](https://github.com/yonaskolb/XcodeGen) types and writes them out to
+disk with [xcodeproj](https://github.com/tuist/XcodeProj).
+
+Generating an Xcode project with XCHammer:
+```
+command line -> bazel -> rule -> xchammer -> xcodegen -> project.xcodeproj
+```
+
+Generators like XCHammer and Tulsi take care of integrating Bazel into the IDE.
+Bazel builds are invoked a shell script build phase from the IDE. Basically,
+Xcode shells out to Bazel to produce the application, and then Xcode picks up
+the product from the derived data path. Performing a Bazel build from Xcode.
+
+```
+play button -> shell script build phase -> bazel build ios-app
+```
+
+![Docs](XcodeBazelRunscript.png)
+
+_In the Xcode world, Xcode's internal build systems produce the application._
+
+## More information
+
+### Gathering information with Bazel query
+
+Bazel query is command line interface that uses a DSL designed to query the
+build graph and extract information. This can be used for tooling purposes, or
+to simply questions about the build. Here's a few examples:
+
+What extensions does the target `pinterest` depend upon?
+```
+bazel query 'kind(ios_extension, deps(pinterest))'
+```
+
+What are the labels of source files in `app-lib`?
+```
+bazel query --noimplicit_deps 'labels(srcs, app-lib)'
+```
+
+Please see the [Bazel query
+documentation](https://docs.bazel.build/versions/master/query.html) for more
+information.
+
+### Fixing common Bazel errors
```

Comment:
Anecdotally, I think this is what most folks find hard about Bazel.  You've got one good example here, but I expect there are other pretty common ones:

1. Attempt to use a rule without `load`ing it first. Someone making a BUILD file for the first time will probably be partially copying an existing one and modifying it.
2. Passing the wrong type to a parameter.
3. Using a target that isn't accessible due to its visibility.

Reply:
To add onto **_1_**, attempting to use a rule without `load`ing it first, and then also forgetting to add the specific method from the .bzl file.